File: javascript/extract-tiktok-yt-dlp.js
#!/usr/bin/env node
// extract-tiktok-yt-dlp.sh — portiert nach javascript
// Quelle: shell, OpenClaw@gateway2:skills/tiktok-live-mon/scripts/extraction-methods/extract-tiktok-yt-dlp.sh
// Erzeugt: 2026-08-09 durch ABSTRACTIONS_MANAGER.py

import { spawnSync } from 'child_process';
import { mkdtempSync, readFileSync, writeFileSync, rmSync } from 'fs';
import { join } from 'path';
import { tmpdir } from 'os';

const USERNAME = process.argv[2]?.replace(/^@/, '') || '';
const FORMAT = process.argv[3] || 'best';
const JSON_FLAG = process.argv[4] || '';
const TIMESTAMP = new Date().toISOString().replace(/\.\d+Z$/, 'Z');

const TMP_DIR = mkdtempSync(join(tmpdir(), 'tiktok-yt-dlp-'));
process.on('exit', () => {
  rmSync(TMP_DIR, { recursive: true, force: true });
});

function emitJson(success, method, username, url, format, error, timestamp, status) {
  const payload = {
    ...(success !== undefined && success !== '' && { success: String(success).toLowerCase() === 'true' }),
    ...(method !== undefined && method !== '' && { method }),
    ...(username !== undefined && username !== '' && { username }),
    ...(url !== undefined && url !== '' && { url }),
    ...(format !== undefined && format !== '' && { format }),
    ...(error !== undefined && error !== '' && { error }),
    ...(timestamp !== undefined && timestamp !== '' && { timestamp }),
    ...(status !== undefined && status !== '' && { status })
  };
  console.error(JSON.stringify(payload, null, 0));
}

if (!USERNAME.match(/^[A-Za-z0-9._]{1,24}$/)) {
  console.error('Invalid TikTok username');
  process.exit(64);
}

const validFormats = [
  'hls-origin/hls-pull/hls-uhd_60/hls-hd_60/hls-hd/hls-sd/hls-ld/flv-origin/flv-hd/flv-ld',
  'hls-uhd_60/hls-hd_60/hls-hd/hls-sd/hls-ld/flv-hd/flv-ld',
  'hls-hd_60/hls-hd/hls-sd/hls-ld/flv-hd/flv-ld',
  'hls-hd/hls-sd/hls-ld/flv-hd/flv-sd/flv-ld',
  'hls-sd/hls-ld/flv-sd/flv-ld',
  'hls-ld/flv-ld',
  'hls-origin/hls-hd/hls-sd/hls-ld/hls-pull/flv-origin/flv-hd/flv-ld'
];

if (!validFormats.includes(FORMAT)) {
  console.error('Invalid yt-dlp format');
  process.exit(64);
}

const loadPerCpu = parseFloat(spawnSync('python3', ['-c', 'import os; print(os.getloadavg()[0] / max(1, os.cpu_count() or 1))']).stdout.toString().trim());
const maxLoad = parseFloat(process.env.TIKTOK_MAX_LOAD_PER_CPU || '1.5');

if (loadPerCpu > maxLoad) {
  emitJson(false, 'yt-dlp', USERNAME, '', FORMAT, 'host overloaded', TIMESTAMP, 'overloaded');
  process.exit(75);
}

if (!spawnSync('command', ['-v', 'yt-dlp']).status === 0) {
  emitJson(false, 'yt-dlp', USERNAME, '', FORMAT, 'yt-dlp not installed', TIMESTAMP, 'dependency_missing');
  process.exit(2);
}

const LIVE_URL = `https://www.tiktok.com/@${USERNAME}/live`;
const stdoutFile = join(TMP_DIR, 'stdout.json');
const stderrFile = join(TMP_DIR, 'stderr.log');

const ytDlpResult = spawnSync('yt-dlp', [
  '--no-warnings',
  '--dump-single-json',
  '--skip-download',
  '--format', FORMAT,
  LIVE_URL
], {
  stdio: ['ignore', 'pipe', 'pipe']
});

writeFileSync(stdoutFile, ytDlpResult.stdout);
writeFileSync(stderrFile, ytDlpResult.stderr);

const EXIT_CODE = ytDlpResult.status;

if (EXIT_CODE !== 0) {
  const stderrContent = readFileSync(stderrFile, 'utf8');
  let STATUS, CODE;
  
  if (/not currently live|No live cdn found|not available|private video/i.test(stderrContent)) {
    STATUS = 'offline';
    CODE = 1;
  } else {
    STATUS = 'technical_error';
    CODE = 2;
  }
  
  emitJson(false, 'yt-dlp', USERNAME, '', FORMAT, stderrContent.substring(0, 1000), TIMESTAMP, STATUS);
  process.exit(CODE);
}

const jsonData = JSON.parse(readFileSync(stdoutFile, 'utf8'));
let URL = '';

if (typeof jsonData.url === 'string') {
  URL = jsonData.url;
} else {
  const formats = jsonData.formats || [];
  for (const item of formats) {
    if (typeof item === 'object' && typeof item.url === 'string') {
      const value = item.url;
      const low = value.toLowerCase();
      if (value.startsWith('https://') && (low.includes('.m3u8') || low.includes('.flv')) && !low.includes('only_audio=1')) {
        URL = value;
        break;
      }
    }
  }
}

if (!URL) {
  emitJson(false, 'yt-dlp', USERNAME, '', FORMAT, 'could not extract HTTPS video URL', TIMESTAMP, 'offline');
  process.exit(1);
}

if (JSON_FLAG === '--json') {
  emitJson(true, 'yt-dlp', USERNAME, URL, FORMAT, '', TIMESTAMP, 'live');
} else {
  console.log(URL);
}

File: python/extract-tiktok-yt-dlp.py
#!/usr/bin/env python3
# extract-tiktok-yt-dlp.sh — portiert nach python
# Quelle: shell, OpenClaw@gateway2:skills/tiktok-live-mon/scripts/extraction-methods/extract-tiktok-yt-dlp.sh
# Erzeugt: 2026-08-09 durch ABSTRACTIONS_MANAGER.py

import json
import os
import re
import subprocess
import sys
import tempfile
from datetime import datetime

def emit_json(*args):
    keys = ("success", "method", "username", "url", "format", "error", "timestamp", "status")
    payload = {k: v for k, v in zip(keys, args) if v != ""}
    if "success" in payload:
        payload["success"] = payload["success"].lower() == "true"
    print(json.dumps(payload, ensure_ascii=False))

def get_load_per_cpu():
    try:
        with open("/proc/loadavg", "r") as f:
            load_avg = f.readline().split()[0]
        load = float(load_avg)
        cpu_count = os.cpu_count() or 1
        return load / max(1, cpu_count)
    except:
        return 0.0

def main():
    if len(sys.argv) < 2:
        print("Invalid TikTok username", file=sys.stderr)
        sys.exit(64)
    
    username = sys.argv[1].lstrip('@')
    format_choice = sys.argv[2] if len(sys.argv) > 2 else "best"
    json_flag = sys.argv[3] if len(sys.argv) > 3 else ""
    
    timestamp = datetime.utcnow().strftime('%Y-%m-%dT%H:%M:%SZ')
    
    valid_username_pattern = r'^[A-Za-z0-9._]{1,24}$'
    if not re.match(valid_username_pattern, username):
        print("Invalid TikTok username", file=sys.stderr)
        sys.exit(64)
    
    valid_formats = [
        "hls-origin/hls-pull/hls-uhd_60/hls-hd_60/hls-hd/hls-sd/hls-ld/flv-origin/flv-hd/flv-ld",
        "hls-uhd_60/hls-hd_60/hls-hd/hls-sd/hls-ld/flv-hd/flv-ld",
        "hls-hd_60/hls-hd/hls-sd/hls-ld/flv-hd/flv-ld",
        "hls-hd/hls-sd/hls-ld/flv-hd/flv-sd/flv-ld",
        "hls-sd/hls-ld/flv-sd/flv-ld",
        "hls-ld/flv-ld",
        "hls-origin/hls-hd/hls-sd/hls-ld/hls-pull/flv-origin/flv-hd/flv-ld"
    ]
    
    if format_choice not in valid_formats:
        print("Invalid yt-dlp format", file=sys.stderr)
        sys.exit(64)
    
    load_per_cpu = get_load_per_cpu()
    max_load = float(os.environ.get("TIKTOK_MAX_LOAD_PER_CPU", "1.5"))
    
    if load_per_cpu > max_load:
        emit_json("false", "yt-dlp", username, "", format_choice, "host overloaded", timestamp, "overloaded")
        sys.exit(75)
    
    try:
        subprocess.run(["yt-dlp", "--version"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, check=True)
    except (subprocess.CalledProcessError, FileNotFoundError):
        emit_json("false", "yt-dlp", username, "", format_choice, "yt-dlp not installed", timestamp, "dependency_missing")
        sys.exit(2)
    
    with tempfile.TemporaryDirectory() as tmp_dir:
        stdout_file = os.path.join(tmp_dir, "stdout.json")
        stderr_file = os.path.join(tmp_dir, "stderr.log")
        
        live_url = f"https://www.tiktok.com/@{username}/live"
        
        try:
            result = subprocess.run([
                "yt-dlp",
                "--no-warnings",
                "--dump-single-json",
                "--skip-download",
                "--format", format_choice,
                live_url
            ], stdout=open(stdout_file, "w"), stderr=open(stderr_file, "w"))
        except Exception as e:
            emit_json("false", "yt-dlp", username, "", format_choice, str(e), timestamp, "technical_error")
            sys.exit(2)
        
        exit_code = result.returncode
        if exit_code != 0:
            try:
                with open(stderr_file, "r") as f:
                    stderr_content = f.read()[:1000]
            except:
                stderr_content = ""
            
            if any(keyword in stderr_content.lower() for keyword in ["not currently live", "no live cdn found", "not available", "private video"]):
                status = "offline"
                code = 1
            else:
                status = "technical_error"
                code = 2
            
            emit_json("false", "yt-dlp", username, "", format_choice, stderr_content, timestamp, status)
            sys.exit(code)
        
        try:
            with open(stdout_file, "r") as f:
                data = json.load(f)
        except Exception as e:
            emit_json("false", "yt-dlp", username, "", format_choice, "could not parse JSON output", timestamp, "technical_error")
            sys.exit(2)
        
        candidates = []
        if isinstance(data.get("url"), str):
            candidates.append(data["url"])
        
        for item in data.get("formats", []) or []:
            if isinstance(item, dict) and isinstance(item.get("url"), str):
                candidates.append(item["url"])
        
        url = ""
        for value in candidates:
            low = value.lower()
            if value.startswith("https://") and (".m3u8" in low or ".flv" in low) and "only_audio=1" not in low:
                url = value
                break
        
        if not url:
            emit_json("false", "yt-dlp", username, "", format_choice, "could not extract HTTPS video URL", timestamp, "offline")
            sys.exit(1)
        
        if json_flag == "--json":
            emit_json("true", "yt-dlp", username, url, format_choice, "", timestamp, "live")
        else:
            print(url)

if __name__ == "__main__":
    main()
